File: client/lib/protocol.dart
import 'dart:typed_data';

/// Frame type constants.
class FrameType {
  static const int connect = 0x01;
  static const int data = 0x02;
  static const int disconnect = 0x03;
  static const int ack = 0x04;
}

const int frameHeaderSize = 5;
const int maxPayloadSize = 32 * 1024;

/// A protocol frame.
class Frame {
  final int type;
  final int channelId;
  final Uint8List payload;

  Frame({required this.type, required this.channelId, required this.payload});

  /// Encode this frame to bytes.
  Uint8List encode() {
    if (payload.length > maxPayloadSize) {
      throw ArgumentError('Payload too large: ${payload.length} > $maxPayloadSize');
    }
    final buf = ByteData(frameHeaderSize + payload.length);
    buf.setUint8(0, type);
    buf.setUint16(1, channelId, Endian.big);
    buf.setUint16(3, payload.length, Endian.big);
    final bytes = buf.buffer.asUint8List();
    bytes.setRange(frameHeaderSize, frameHeaderSize + payload.length, payload);
    return bytes;
  }

  /// Decode a single frame from bytes.
  static Frame decode(Uint8List data) {
    if (data.length < frameHeaderSize) {
      throw FormatException('Frame too short: ${data.length}');
    }
    final bd = ByteData.sublistView(data);
    final type = bd.getUint8(0);
    final channelId = bd.getUint16(1, Endian.big);
    final payloadLen = bd.getUint16(3, Endian.big);
    if (payloadLen > maxPayloadSize) {
      throw FormatException('Payload length $payloadLen exceeds max $maxPayloadSize');
    }
    if (data.length < frameHeaderSize + payloadLen) {
      throw FormatException(
          'Frame truncated: need ${frameHeaderSize + payloadLen}, got ${data.length}');
    }
    final payload = Uint8List.fromList(
        data.sublist(frameHeaderSize, frameHeaderSize + payloadLen));
    return Frame(type: type, channelId: channelId, payload: payload);
  }

  /// Decode multiple concatenated frames.
  static List<Frame> decodeAll(Uint8List data) {
    final frames = <Frame>[];
    var offset = 0;
    while (offset < data.length) {
      if (data.length - offset < frameHeaderSize) {
        throw FormatException('Trailing bytes too short for frame header');
      }
      final bd = ByteData.sublistView(data, offset);
      final payloadLen = bd.getUint16(3, Endian.big);
      final frameLen = frameHeaderSize + payloadLen;
      if (offset + frameLen > data.length) {
        throw FormatException('Frame extends past data boundary');
      }
      frames.add(Frame.decode(Uint8List.fromList(
          data.sublist(offset, offset + frameLen))));
      offset += frameLen;
    }
    return frames;
  }
}

File: server/protocol.go
package main

import (
	"encoding/binary"
	"errors"
	"fmt"
)

// Frame types
const (
	FrameConnect    byte = 0x01
	FrameData       byte = 0x02
	FrameDisconnect byte = 0x03
	FrameAck        byte = 0x04
)

// Frame header size: 1 (type) + 2 (channel_id) + 2 (payload_length)
const frameHeaderSize = 5

// MaxPayloadSize limits individual frame payload.
const MaxPayloadSize = 32 * 1024

// Frame represents a protocol frame.
type Frame struct {
	Type      byte
	ChannelID uint16
	Payload   []byte
}

// EncodeFrame serialises a Frame into bytes.
func EncodeFrame(f Frame) ([]byte, error) {
	if len(f.Payload) > MaxPayloadSize {
		return nil, fmt.Errorf("payload too large: %d > %d", len(f.Payload), MaxPayloadSize)
	}
	buf := make([]byte, frameHeaderSize+len(f.Payload))
	buf[0] = f.Type
	binary.BigEndian.PutUint16(buf[1:3], f.ChannelID)
	binary.BigEndian.PutUint16(buf[3:5], uint16(len(f.Payload)))
	copy(buf[5:], f.Payload)
	return buf, nil
}

// DecodeFrame deserialises bytes into a Frame.
func DecodeFrame(data []byte) (Frame, error) {
	if len(data) < frameHeaderSize {
		return Frame{}, errors.New("frame too short")
	}
	f := Frame{
		Type:      data[0],
		ChannelID: binary.BigEndian.Uint16(data[1:3]),
	}
	payloadLen := int(binary.BigEndian.Uint16(data[3:5]))
	if payloadLen > MaxPayloadSize {
		return Frame{}, fmt.Errorf("payload length %d exceeds max %d", payloadLen, MaxPayloadSize)
	}
	if len(data) < frameHeaderSize+payloadLen {
		return Frame{}, fmt.Errorf("frame truncated: need %d, got %d", frameHeaderSize+payloadLen, len(data))
	}
	f.Payload = make([]byte, payloadLen)
	copy(f.Payload, data[frameHeaderSize:frameHeaderSize+payloadLen])
	return f, nil
}

// DecodeFrames decodes multiple concatenated frames from a byte slice.
func DecodeFrames(data []byte) ([]Frame, error) {
	var frames []Frame
	offset := 0
	for offset < len(data) {
		if len(data)-offset < frameHeaderSize {
			return nil, errors.New("trailing bytes too short for frame header")
		}
		payloadLen := int(binary.BigEndian.Uint16(data[offset+3 : offset+5]))
		frameLen := frameHeaderSize + payloadLen
		if offset+frameLen > len(data) {
			return nil, errors.New("frame extends past data boundary")
		}
		f, err := DecodeFrame(data[offset : offset+frameLen])
		if err != nil {
			return nil, err
		}
		frames = append(frames, f)
		offset += frameLen
	}
	return frames, nil
}
